src/web_app/js/background.js

```diff
@@ -9,14 +9,151 @@
 /* More information about these options at jshint.com/docs/options */
 
 // Variables defined in and used from chrome.
-/* globals chrome */
+/* globals chrome, sendAsyncUrlRequest, Constants, parseJSON */
 
 'use strict';
-chrome.app.runtime.onLaunched.addListener(function() {
-  chrome.app.window.create('appwindow.html', {
-    'width': 800,
-    'height': 600,
-    'left': 0,
-    'top': 0
+(function() {
+  chrome.app.runtime.onLaunched.addListener(function() {
+    chrome.app.window.create('appwindow.html', {
+      'width': 800,
+      'height': 600,
+      'left': 0,
+      'top': 0
+    });
   });
-});
+
+  // Send event notification from background to window.
+  var sendWSEventMessageToWindow = function(port, wsEvent, data) {
+    var message = {
+        action: Constants.WS_ACTION,
+        wsAction: Constants.EVENT_ACTION,
+        wsEvent: wsEvent
+    };
+    if (data) {
+      message.data = data;
+    }
+    trace('B -> W: ' + JSON.stringify(message));
+    try {
+      port.postMessage(message);
+    } catch (ex) {
+      trace('Error sending message: ' + ex);
+    }
+  };
+
+  var handleWebSocketRequest = function(port, message) {
+    if (message.wsAction === Constants.WS_CREATE_ACTION) {
+      trace('RWS: creating web socket: ' + message.wssUrl);
+      var ws = new WebSocket(message.wssUrl);
+      port.wssPostUrl_ = message.wssPostUrl;
+      ws.onopen = function() {
+        trace('RWS: onopen');
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONOPEN);
+      };
+      ws.onerror = function() {
+        trace('RWS: onerror.');
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONERROR);
+      };
+      ws.onclose = function(event) {
+        trace('RWS: onclose - code: ' + event.code + ' reason: ' +
+            event.reason);
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONCLOSE, event);
+      };
+      ws.onmessage = function(event) {
+        trace('RWS: onmessage: ' + event.data);
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONMESSAGE, event);
+      };
+      port.webSocket_ = ws;
+    } else if (message.wsAction === Constants.WS_SEND_ACTION) {
+      trace('RWS: sending: ' + message.data);
+      if (port.webSocket_ && port.webSocket_.readyState === WebSocket.OPEN) {
+        try {
+          port.webSocket_.send(message.data);
+        } catch (ex) {
+          trace('RWS: exception sending: ' + ex);
+        }
+      } else {
+        // Attempt to send message using wss port url.
+        trace('RWS: web socket not ready, falling back to POST.');
+        var msg = parseJSON(message.data);
+        if (msg) {
+          sendAsyncUrlRequest('POST', port.wssPostUrl_, msg.msg);
+        }
+      }
+    } else if (message.wsAction === Constants.WS_CLOSE_ACTION) {
+      trace('RWS: close');
+      port.webSocket_.close();
+      port.webSocket = null;
+    }
+  };
+
+  var createCleanupTaskPromise = function(port, message) {
```

Comment:
If you call the inner function executeCleanupTask(port, message), you can just make the caller do 
promise.then(executeCleanupTask.bind(port, message))